File: geo_engine/rules/links.py
"""Reglas de enlaces: target/rel obligatorios y proxy eLibro (Reglas 14 y 15)."""
from __future__ import annotations

import re
from typing import Any, Dict, List, Tuple

from ..findings import Finding, SEVERITY_WARNING
from ..utils import line_at, snippet_at
from .base import Rule

_A_RE = re.compile(r"<a\b([^>]*)>", re.IGNORECASE)
_HREF_RE = re.compile(r'href\s*=\s*"(.*?)"', re.IGNORECASE)
_TARGET_RE = re.compile(r'target\s*=', re.IGNORECASE)
_REL_RE = re.compile(r'rel\s*=', re.IGNORECASE)


class LinkTargetRule(Rule):
    id = "link-target"
    description = 'Los <a> (salvo ancla #) deben tener target="_blank" y rel.'
    severity = SEVERITY_WARNING
    auto_fixable = True

    def _missing(self, attrs: str):
        """Devuelve (skip, falta_target, falta_rel) para los atributos de un <a>."""
        href = _HREF_RE.search(attrs)
        if href is not None and href.group(1).strip().startswith("#"):
            return True, False, False
        return False, _TARGET_RE.search(attrs) is None, _REL_RE.search(attrs) is None

    def check(self, html: str, ctx: Dict[str, Any]) -> List[Finding]:
        out: List[Finding] = []
        for m in _A_RE.finditer(html):
            skip, no_target, no_rel = self._missing(m.group(1))
            if skip or not (no_target or no_rel):
                continue
            faltan = []
            if no_target:
                faltan.append('target="_blank"')
            if no_rel:
                faltan.append("rel")
            out.append(Finding(
                self.id, self.severity,
                "Enlace sin %s." % " y ".join(faltan),
                line_at(html, m.start()), snippet_at(html, m.start(), m.end()),
            ))
        return out

    def fix(self, html: str, ctx: Dict[str, Any]) -> Tuple[str, List[Finding]]:
        rel_value = self.options.get("rel", "noopener")
        fixed: List[Finding] = []

        def _sub(m: "re.Match[str]") -> str:
            attrs = m.group(1)
            skip, no_target, no_rel = self._missing(attrs)
            if skip or not (no_target or no_rel):
                return m.group(0)
            additions = ""
            if no_target:
                additions += ' target="_blank"'
            if no_rel:
                additions += ' rel="%s"' % rel_value
            fixed.append(Finding(
                self.id, self.severity,
                "Añadido%s al enlace." % additions,
                line_at(html, m.start()), fixed=True,
            ))
            return "<a" + attrs.rstrip() + additions + ">"

        return _A_RE.sub(_sub, html), fixed


class ElibroProxyRule(Rule):
    id = "elibro-proxy"
    description = "El proxy eLibro debe llevar guion: elibro-net.ezproxy.udes.edu.co."
    severity = SEVERITY_WARNING
    auto_fixable = True

    _BAD = "elibronet.ezproxy.udes.edu.co"
    _GOOD = "elibro-net.ezproxy.udes.edu.co"

    def check(self, html: str, ctx: Dict[str, Any]) -> List[Finding]:
        out: List[Finding] = []
        start = 0
        while True:
            idx = html.find(self._BAD, start)
            if idx == -1:
                break
            out.append(Finding(
                self.id, self.severity,
                "Proxy eLibro sin guion (fallará en la red institucional).",
                line_at(html, idx), snippet_at(html, idx, idx + len(self._BAD)),
            ))
            start = idx + len(self._BAD)
        return out

    def fix(self, html: str, ctx: Dict[str, Any]) -> Tuple[str, List[Finding]]:
        if self._BAD not in html:
            return html, []
        count = html.count(self._BAD)
        fixed = [Finding(
            self.id, self.severity,
            "Corregido proxy eLibro (%d ocurrencia(s))." % count, fixed=True,
        )]
        return html.replace(self._BAD, self._GOOD), fixed

File: web/src/linter/rules/links.js
/**
 * @module rules/links
 * Link-related lint rules (CHECK-ONLY).
 * Ported from Python links.py.
 *
 * Rules:
 *   - LinkTargetRule       (link-target)
 *   - ElibroProxyRule      (elibro-proxy)
 *   - ForbiddenSourceRule  (forbidden-source)
 *   - PluginfileRedRule    (pluginfile-red)
 */

import { Finding, SEVERITY_ERROR, SEVERITY_WARNING, lineAt, snippetAt } from '../Finding.js';

// ─── LinkTargetRule ─────────────────────────────────────────────────────────

/**
 * Ensures external `<a>` tags have `target="_blank"` and a `rel` attribute.
 * Anchors whose `href` starts with `#` (in-page links) are skipped.
 */
export class LinkTargetRule {
  static id = 'link-target';
  static description = 'External link missing target="_blank" or rel attribute';
  static severity = SEVERITY_WARNING;

  constructor(options = {}) {
    this.options = options;
  }

  /**
   * @param {string} html
   * @returns {Finding[]}
   */
  check(html) {
    const findings = [];
    const tagPattern = /<a\b([^>]*)>/gi;
    const hrefRe = /href\s*=\s*"(.*?)"/i;
    const targetRe = /target\s*=/i;
    const relRe = /rel\s*=/i;

    let m;
    while ((m = tagPattern.exec(html)) !== null) {
      const attrs = m[1];
      const hrefMatch = hrefRe.exec(attrs);
      const href = hrefMatch ? hrefMatch[1] : '';

      // Skip in-page anchors.
      if (href.startsWith('#')) continue;

      const hasTarget = targetRe.test(attrs);
      const hasRel = relRe.test(attrs);
      const issues = [];

      if (!hasTarget) issues.push('target="_blank"');
      if (!hasRel) issues.push('rel attribute');

      if (issues.length > 0) {
        findings.push(
          new Finding({
            ruleId: LinkTargetRule.id,
            severity: LinkTargetRule.severity,
            message: `Link missing ${issues.join(' and ')}`,
            line: lineAt(html, m.index),
            snippet: snippetAt(html, m.index, m.index + m[0].length),
          }),
        );
      }
    }
    return findings;
  }
}

// ─── ElibroProxyRule ────────────────────────────────────────────────────────

/**
 * Detects the incorrect eLibro proxy domain (missing hyphen).
 * Bad:  `elibronet.ezproxy.udes.edu.co`
 * Good: `elibro-net.ezproxy.udes.edu.co`
 */
export class ElibroProxyRule {
  static id = 'elibro-proxy';
  static description = 'Incorrect eLibro proxy URL (missing hyphen)';
  static severity = SEVERITY_ERROR;

  constructor(options = {}) {
    this.options = options;
  }

  /**
   * @param {string} html
   * @returns {Finding[]}
   */
  check(html) {
    const findings = [];
    const BAD = 'elibronet.ezproxy.udes.edu.co';
    let pos = 0;

    while (true) {
      const idx = html.indexOf(BAD, pos);
      if (idx === -1) break;

      findings.push(
        new Finding({
          ruleId: ElibroProxyRule.id,
          severity: ElibroProxyRule.severity,
          message:
            'Incorrect eLibro proxy URL — should be elibro-net.ezproxy.udes.edu.co',
          line: lineAt(html, idx),
          snippet: snippetAt(html, idx, idx + BAD.length),
        }),
      );
      pos = idx + BAD.length;
    }
    return findings;
  }
}

// ─── ForbiddenSourceRule ────────────────────────────────────────────────────

/**
 * Detects forbidden file sources (regla GEO: los archivos locales van con
 * `@@PLUGINFILE@@`; nunca borradores de Moodle ni nubes externas):
 *   - `draftfile.php`         — borrador temporal de Moodle, el enlace muere.
 *   - OneDrive / SharePoint   — requiere cuenta externa, rompe la trazabilidad.
 */
export class ForbiddenSourceRule {
  static id = 'forbidden-source';
  static description = 'Forbidden file source (draftfile.php / OneDrive / SharePoint)';
  static severity = SEVERITY_ERROR;

  constructor(options = {}) {
    this.options = options;
  }

  /**
   * @param {string} html
   * @returns {Finding[]}
   */
  check(html) {
    const findings = [];
    const SOURCES = [
      { re: /draftfile\.php/gi, msg: 'Enlace a draftfile.php (borrador temporal) — usa @@PLUGINFILE@@' },
      { re: /(?:1drv\.ms|onedrive\.live\.com|sharepoint\.com|onedrive\.com)/gi, msg: 'Enlace a OneDrive/SharePoint (prohibido) — sube el archivo y usa @@PLUGINFILE@@' },
    ];

    for (const { re, msg } of SOURCES) {
      let m;
      while ((m = re.exec(html)) !== null) {
        findings.push(
          new Finding({
            ruleId: ForbiddenSourceRule.id,
            severity: ForbiddenSourceRule.severity,
            message: msg,
            line: lineAt(html, m.index),
            snippet: snippetAt(html, Math.max(0, m.index - 30), m.index + m[0].length + 30),
          }),
        );
      }
    }
    return findings;
  }
}

// ─── PluginfileRedRule ──────────────────────────────────────────────────────

/**
 * Cross-checks every `@@PLUGINFILE@@/<archivo>` reference against the list of
 * RED filenames registered in the wizard.  Only runs when the list is
 * non-empty (without it there is nothing to compare against).
 *
 * The `redFiles` option is read live on each check, so the app can mutate the
 * same array as the user registers files.
 */
export class PluginfileRedRule {
  static id = 'pluginfile-red';
  static description = '@@PLUGINFILE@@ reference does not match any registered RED file';
  static severity = SEVERITY_WARNING;

  /**
   * @param {object} [options]
   * @param {string[]} [options.redFiles=[]] - Known RED filenames (live reference).
   */
  constructor(options = {}) {
    this.options = options;
    this.redFiles = options.redFiles ?? [];
  }

  /** Normalise a filename for comparison. @param {string} name */
  static norm(name) {
    let n = name;
    try { n = decodeURIComponent(n); } catch { /* keep raw */ }
    return n.replace(/\s+/g, ' ').trim().toLowerCase();
  }

  /**
   * @param {string} html
   * @returns {Finding[]}
   */
  check(html) {
    const findings = [];
    if (!this.redFiles || this.redFiles.length === 0) return findings;

    const known = new Set(this.redFiles.map(PluginfileRedRule.norm));
    const pattern = /@@PLUGINFILE@@\/([^"'<>\s?#]+)/g;
    let m;
    while ((m = pattern.exec(html)) !== null) {
      const name = PluginfileRedRule.norm(m[1]);
      if (known.has(name)) continue;

      findings.push(
        new Finding({
          ruleId: PluginfileRedRule.id,
          severity: PluginfileRedRule.severity,
          message: `"${m[1]}" no coincide con ningún archivo RED registrado — revisa el nombre exacto`,
          line: lineAt(html, m.index),
          snippet: snippetAt(html, m.index, m.index + m[0].length),
        }),
      );
    }
    return findings;
  }
}
